### PR TITLE
Use a precompiled NIF

### PR DESCRIPTION
**Pull request by philss**
This PR introduces the ability to download a precompiled NIF and remove the need of having the Rust toolchain installed.

It is going to work for the majority of popular Operational Systems.
The following [targets](https://doc.rust-lang.org/cargo/appendix/glossary.html#target) are supported:
- aarch64-apple-darwin - MacOS on M1
- x86_64-apple-darwin
- x86_64-unknown-linux-gnu
- x86_64-unknown-linux-musl 
- arm-unknown-linux-gnueabihf
- x86_64-pc-windows-gnu
- x86_64-pc-windows-msvc

You can play with it using this snippet (requires Elixir ~> 1.12):

```elixir
Mix.install([
  {:html5ever,
   git: "https://github.com/philss/html5ever_elixir.git", branch: "ps-download-precompiled"}
])

html = """
<!doctype html>
<html>
  <head><title>Testing</title></head>
  <body><h1>Hello world</h1></body>
</html>
"""

Html5ever.parse(html)
```

## The Release workflow

Most of the work is done by the CI server, which in this case is GitHub Actions.
Upon each new pushed tag we trigger the "release" workflow - which is independent of Hex.
That workflow is going to build the NIF files and upload to the release page. We use the tool ["cross"](https://github.com/rust-embedded/cross) to help with some of the cross compilations.

You can see one of the releases here: https://github.com/philss/html5ever_elixir/releases/tag/testing-release30

The "release" workflow was heavily inspired by this: https://github.com/sharkdp/lscolors/blob/c6e191b91f637058ba932a4f7b065f2e19d8a1ba/.github/workflows/CICD.yml

## Open questions

- Should we compress the files (tar.gz)?
- Should we include the license files together?
- In case the precompiled NIF is not available, should we always force a compilation? Today we raise an error.

### CODE REVIEW

File: native/html5ever_nif/Cargo.toml
Comment (philss) [inline native/html5ever_nif/Cargo.toml:13]: We should wait for a new release of `rustler` since the current version (0.22) did not work on Windows (It's probably the OTP that is not installed).

File: native/html5ever_nif/Cargo.toml
Comment (philss) [inline native/html5ever_nif/Cargo.toml:10]: This is the crate type recommended for shared libs that are not going to be used in Rust programs. See https://doc.rust-lang.org/reference/linkage.html

File: lib/html5ever/native.ex
Comment (josevalim) [inline lib/html5ever/native.ex:29]: Why the OS check? :)

File: lib/html5ever/precompiled.ex
Comment (josevalim) [inline lib/html5ever/precompiled.ex:66]: I don't think we should use Mix target. The Mix target can be named anything, those names are only conveniences. Instead you should pattern match on `system_arch` instead, like we do here: https://github.com/phoenixframework/esbuild/blob/0272743c3e1d5e4898a2e24c20276b4cf0370180/lib/esbuild.ex#L220-L242

File: mix.exs
Comment (josevalim) [inline mix.exs:29]:
```suggestion
      {:castore, "~> 0.1"},
```

File: lib/html5ever/precompiled.ex
Comment (josevalim) [inline lib/html5ever/precompiled.ex:112]: For XLA/Torch, we are caching at `Path.join(File.cwd!(), "cache")`, instead of using a system cache.

  * Pros: we don't write outside of the project directory, this may also be easier to cache on CI?
  * Cons: each different project has to download it from scratch.

WDYT? @wojtekmach, any preferences?

Comment (wojtekmach) [reply lib/html5ever/precompiled.ex:112]: Not a strong opinion but so think the global cache would be nice to avoid duplication. We can use :filename.basedir with :user_cache argument. And if someone has `MIX_ENV` var set, we pass `os: :Linux` to use XDG. Wdyt?

Comment (josevalim) [reply lib/html5ever/precompiled.ex:112]: XLA stores it on the user_cache too, so this is good as is. In the future we can make the path customizable too, as in XLA: https://github.com/elixir-nx/xla/blob/main/lib/xla.ex#L59-L66

File: lib/html5ever/native.ex
Comment (philss) [reply lib/html5ever/native.ex:29]: I wasn't raising the error for Windows because I didn't have the files. I can remove this now :)

File: lib/html5ever/precompiled.ex
Comment (philss) [reply lib/html5ever/precompiled.ex:66]: The target here is only for use cases like Nerves. They set the `MIX_TARGET` env var to configure the "cross compilation" to Nerves. If I don't do this, then my host sets the `system_arch` and we download the wrong NIF. I will double check if this is needed.

Comment (josevalim) [reply lib/html5ever/precompiled.ex:66]: Hrm... let's check with the nerves team. It may be other env variables to check.

Comment (philss) [reply lib/html5ever/precompiled.ex]: I managed to make this work with the help of Frank :)
It's now using the `TARGET_*` env vars.

File: lib/html5ever/precompiled.ex
Comment (josevalim) [inline lib/html5ever/precompiled.ex:155]:
```suggestion
    priv_dir = Application.app_dir(name, "priv")
```

**Review by josevalim — APPROVED** [review]
Absolutely beautiful. Just one minor comment and ship it!

File: lib/html5ever/precompiled.ex
Comment (josevalim) [inline lib/html5ever/precompiled.ex:61]: Please include the supported nif versions in the error message.

File: lib/html5ever/precompiled.ex
Comment (josevalim) [inline lib/html5ever/precompiled.ex:57]: Please include the supported archs in the error message.

File: lib/html5ever/native.ex
Comment (josevalim) [inline lib/html5ever/native.ex:27]:
```suggestion
          error = "Error while downloading precompiled NIF: #{error}. Set HTML5EVER_BUILD=1 to compile the NIF from scratch"
```

File: lib/html5ever/precompiled.ex
Comment (josevalim) [inline lib/html5ever/precompiled.ex:59]: Hrm, I am thinking a little about this. This has the following issue: if there is a new OTP version with a new nif version, then you cannot update the OTP version your app uses unless we go back in time and release new precompiled packages for all previous versions.

So I am thinking we should pick the more recent compatible NIF version:

    defp find_compatible_nif_version(vsn, available) do
      if vsn in available do
        vsn
      else
        [major | _] = String.split(".")

        available
        |> Enum.filter(fn vsn -> String.starts_with?(vsn, major <> ".") end)
        |> Enum.max_by(fn vsn -> vsn |> String.split(".") |> Enum.map(&String.to_integer/1) end)
    end

**Comment by josevalim** [comment]
@philss btw, I think we should .tar.gz it. Can you do a quick test and see how much size does it save?

**Comment by philss** [comment]
@josevalim the files are much smaller now. I will change the load process to decompress it. 

## Previous
release31

## Current - with tar.gz
release32

PS: the maOS build failed for the current. This does not change the results.

File: lib/html5ever/precompiled.ex
Comment (philss) [reply lib/html5ever/precompiled.ex:70]: OK, I will try that. Thanks!

Comment (philss) [reply lib/html5ever/precompiled.ex:68]: I kept the error message in case we can't find a compatible version (likely when a major change occurs).

**Comment by hansihe** [comment]
Really nice work!

Some thoughts, sorry this is a bit of a mess:

### Open questions

 * > Should we compress the files (tar.gz)?
   
   I think it makes sense to do so. This also makes it possible to package other files in there, but this is probably not needed yet.

 * > In case the precompiled NIF is not available, should we always force a compilation? Today we raise an error.
   
   I think raising is the correct behavior by default. We also need an option somehow to enable compilation.

   An environment variable might not be the right option here, is this something we could put in config instead?

### NIF versions

This is the quote that José sent me from the OTP team regarding NIF version compatibility:

> It’s a bit of both, we guarantee not to make NIFs incompatible for two major OTP versions, and that you can load NIFs of the same major NIF version.
That is, a NIF compiled for OTP 24 will be loadable in OTP 26, and probably even OTP 27 if we don’t bump the NIF major version.

Usually a NIF will use a specific feature set, and that is usally the limiting factor as to what NIF versions it supports. 

If a NIF doesn't use any new features from `2.15` or `2.16`, It could compile for `2.14`. Since this is a minor NIF version bump, it should be compatible with all future BEAM versions supporting `2.xx` NIF versions.

Is there any advantage to compiling versions supporting `2.15` and `2.16` if a NIF compiled for `2.14` could be loaded just fine?

The downloading logic could probably choose the latest compatible NIF version (major version matching, minor version <= current).

Then what NIFs to compile would be a concern for the CI configuration. If no recent features are used, the library might choose to build only one version. If newer NIF API versions provide some advantage (new NIF API function added that makes things more efficient or something), several versions could be compiled.

---

The version selection logic seems to be already implemented, but (unless I misread) it might not be filtering for minor version <= current VM minor version?

File: lib/html5ever/precompiled.ex
Comment (hansihe) [inline lib/html5ever/precompiled.ex:98]: Is this missing a filter for (minor version) <= (minor version of current BEAM)?

File: native/html5ever_nif/Cargo.toml
Comment (hansihe) [reply native/html5ever_nif/Cargo.toml:10]: Nice catch!

**Comment by josevalim** [comment]
> An environment variable might not be the right option here, is this something we could put in config instead?

A configuration is not a bad idea. We could support both if necessary. Keep in mind that env vars can also be given to deps: `{:html5ever, "~> 1.0", system_env: %{"HTML5EVER_BUILD" => 1}}`

> Is there any advantage to compiling versions supporting 2.15 and 2.16 if a NIF compiled for 2.14 could be loaded just fine?

Could we have optimizations/enhancements/fixes in 2.16 that are not leveraged by 2.14? 🤔 I am not well-versed on the C side of things to say.

> The version selection logic seems to be already implemented, but (unless I misread) it might not be filtering for minor version <= current VM minor version?

You are correct. Great catch.

File: lib/html5ever/precompiled.ex
Comment (josevalim) [reply lib/html5ever/precompiled.ex:98]: @philss I think we need something like this:

```
  defp find_compatible_nif_version(vsn, available) do
    if vsn in available do
      {:ok, vsn}
    else
      [major, minor | _] = parse_version(vsn)

      available
      |> Enum.map(&parse_version/1)
      |> Enum.filter(fn
        [^major, available_minor | _] when available_minor >= minor -> true
        [_ | _] -> false
      end)
      |> case do
        [] -> :error
        match -> {:ok, match |> Enum.max() |> Enum.join(".")}
      end
    end
  end

  defp parse_version(vsn) do
    vsn |> String.split(".") |> Enum.map(&String.to_integer/1)
  end
```

We should expose this function (with `@doc false` or on a private module) and write some unit tests.

**Comment by hansihe** [comment]
> A configuration is not a bad idea. We could support both if necessary. Keep in mind that env vars can also be given to deps: `{:html5ever, "~> 1.0", system_env: %{"HTML5EVER_BUILD" => 1}}`

What's your thoughts about this?

I was thinking that whether you want to fall back to compilation is more an application config concern than something you want to set in your environment. I imagine it could get very annoying making sure you have the right shell environment set up in order for mix invocations to work.

I don't necessarily want my libraries to try invoking Rust behind my back by default, but if I explicitly opt into it on an application config level, I want it to be as unannoying as possible to deal with.

> Could we have optimizations/enhancements/fixes in 2.16 that are not leveraged by 2.14? thinking I am not well-versed on the C side of things to say.

I don't mean specifically for `html5ever_elixir`, more that it's something we want to consider.

An example:

Rustler has functions for getting the type of a term (`Term::get_type`). 

Before `OTP 22.0`, this function had to be implemented as calls to a chain of `is_[...]` functions in the NIF API, since there was no function that returned the type of the term.

After `OTP 22.0`, we got `enif_term_type`.

This means that, while `Term::get_type` works on previous versions as well, it can have a much more efficient implementation on newer NIF API versions.

Maybe not the best example, but at least it demonstrates that this might be something we want to think about.

**Comment by josevalim** [comment]
> I was thinking that whether you want to fall back to compilation is more an application config concern than something you want to set in your environment.

I think it can be both. For example, I may be running on a platform that has no precompiled archives, so I want to force the build on my machine but no affect my coworkers. :) I agree on the application environment examples you gave too. 

> I don't mean specifically for html5ever_elixir, more that it's something we want to consider for when some of this work makes it into Rustler.

Ah, really good example. So that's a reason to compile to the latest nif version possible if we can, no?

**Comment by hansihe** [comment]
> Ah, really good example. So that's a reason to compile to the latest nif version possible if we can, no?

Yep, it's probably a good idea to compile for all versions now that I think about it. The always-safe approach is to select the latest possible version, and it's not like we are limited on CI resources :)

**Comment by philss** [comment]
@hansihe thanks for the review!

For the application config I choose to not add a new option because `:skip_compilation?` does something similar. WDYT about using that?
If we choose to add another option, we are probably need to define something outside the NIF module scope, since Rustler uses that to build its own config (we can also change Rustler to accept that).

@josevalim thanks for the suggestion! I added some tests to it. Please check.

File: README.md
Comment (josevalim) [inline README.md:25]: What if we use an option such as `build_from_source: true` to mirror the `_BUILD` env var?

**Review by josevalim — COMMENTED** [review]
Just one tiny comment and then :shipit: :shipit: :shipit: !

File: README.md
Comment (philss) [reply README.md:25]: @josevalim makes sense! I will do that outside the scope of the NIF module because I would have to change Rustler for that.

**Review by hansihe — APPROVED** [review]
Nice work, LGTM!

**Comment by hansihe** [comment]
Feel free to merge whenever you are ready @philss.

We should remember to change the URLs as appropriate before merging.

File: README.md
Comment (josevalim) [inline README.md:25]: Don't you want to keep it as `Html5ever.Native` because that's likely the module we would use if we move this to Rustler? :)

Comment (josevalim) [reply README.md:25]: Or keep it as is and we can change later (when we move it to Rustler).

Comment (philss) [reply README.md:25]: Yeah, I will leave as it is because I need to change Rustler in order to use the same module from the NIF. I will change that in the future.

**Pull request merged**
